src/bin/pg_autoctl/supervisor.h

```diff
@@ -108,19 +119,49 @@ typedef struct Supervisor
 {
 	Service *services;
 	int serviceCount;
+
 	char pidfile[MAXPGPATH];
 	pid_t pid;
 	bool cleanExit;
 	bool shutdownSequenceInProgress;
 	int shutdownSignal;
 	int stoppingLoopCounter;
-} Supervisor;
 
+	/* dynamic handler function */
+	int (*dynamicHandler)(struct Supervisor *, void *);
+	void *dynamicHandlerArg;    /* dynamic handler private data */
 
-bool supervisor_start(Service services[], int serviceCount, const char *pidfile);
+	/*
+	 * Anonymous struct holding active dynamic services.
+	 * Do not access directly.
+	 */
+	struct
+	{
+		Service services[MAXDYNSERVICES];
+		int serviceCount;
+	} dynamicServices;
```

Comment:
I'm not fond of anonymous data structures. Is there a drawback to exposing that as a first-level structure that is contained in the Supervisor structure?

Reply:
I will definitely not hold a strong opinion for this.

The intention is to make as obvious as possible the 'not yours to touch' aspect, just before going to the whole `void *` private method which seems to not be fitting the rest of the code.

Will follow any paradigm you prefer as maintainer of this code :)

Reply:
My favourite paradigm is to user-focus design and ease of maintenance. I believe that our code should be as boring as possible. My experience is that anything interesting you write in C is going to come and bite you later...